### ISSUE TITLE
[BUG] pow operator not acting as expected.

### ISSUE THREAD

**Issue by bdice** [rapidsai/cudf#10178]
**Describe the bug**
`__pow__` / `__rpow__` binary operations do not give the expected results in the following cases.

**Steps/Code to reproduce bug**
Powers of `3 ** 1`.
```python
>>> 3 ** cudf.Series([1])
0    2
dtype: int64
>>> cudf.Series([3]) ** 1
0    2
dtype: int64
>>> pow(cudf.Series([3]), 1)
0    2
dtype: int64
>>> cudf.Series([3]) ** cudf.Series([1])
0    2
dtype: int64
>>> cudf.Series([-3]) ** 1
0   -2
dtype: int64
>>> 3 ** cudf.Series([1, 2, 3, 4])
0     2
1     9
2    27
3    81
dtype: int64
```

This happens for other integers as well:
```python
for i in range(1000):
    if i != (cudf.Series([i]) ** 1)[0]:
        print(i)
```
gives
```
3
84
86
95
117
147
158
213
...
```

Also, the power does not have to be 1. There are a lot of integers that fail, always off-by-one in the result.
```python
>>> print((cudf.Series([8]) ** 7)[0], 8 ** 7)
2097151 2097152
```

It gives the wrong results for Series with dtypes `int8`, `int16`, `int32`, `int64`, `uint8`, `uint16`, `uint32`, `uint64`, `bool` (for the exponent, `3 ** cudf.Series([1], dtype=bool)` gives `2`).

Additionally, pandas raises when computing integers to negative integer powers. cudf gives meaningless results.
```python
>>> cudf.Series([0, 1, 2, 3]) ** -1
0    9223372036854775807
1                      1
2                      0
3                      0
dtype: int64
>>> pd.Series([0, 1, 2, 3]) ** -1
(... traceback ...)
ValueError: Integers to negative integer powers are not allowed.
```

**Expected behavior**
```python
>>> 3 ** 1
3
```

**Additional context**
[Three is a magic number.](https://youtu.be/J8lRKCw2_Pk)

Using a floating type on either side gives expected results:
```python
>>> 3 ** cudf.Series([1, 2, 3, 4], dtype=float)
0     3.0
1     9.0
2    27.0
3    81.0
dtype: float64
```

**Comment by vyasr** [comment]
This appears to be some sort of precision issue that manifests for specific numbers. I can reproduce on my machine, and moreover there are other apparently unrelated numbers conforming to no specific pattern that exhibit the same behavior of `x**1 == x-1`.
```
>>> x = cudf.Series([1, 2]) 
>>> for i in range(100):
...     if (i ** x)[0] != i:
...             print(f"{i}**1 returns {(i**x)[0]}")
... 
3**1 returns 2
84**1 returns 83
86**1 returns 85
95**1 returns 94
```

**Comment by bdice** [comment]
This comment summarizes some Slack discussions, analysis, and plans for fixing this bug.

C++'s `[std::pow](https://en.cppreference.com/w/cpp/numeric/math/pow)` always returns `double`. Current our libcudf implementation of `Pow` always [casts to `double`](https://github.com/rapidsai/cudf/blob/f4ac6d46a04995a43480050adf8d005f25336055/cpp/src/binaryop/compiled/operation.cuh#L215) but then machinery in `[binary_ops.cuh](https://github.com/rapidsai/cudf/blob/f4ac6d46a04995a43480050adf8d005f25336055/cpp/src/binaryop/compiled/binary_ops.cuh#L177)` casts the result to the common type of the inputs, i.e. calling `Pow` on integer types yields a `double` but it is cast back to an integer.

We want the `Pow` struct to conform to STL C++ conventions. NumPy, Pandas, and CuPy all return integer types when computing powers of integer bases to integer exponents. For instance, [cupy handles powers of integers with exponentiation by squaring](https://github.com/cupy/cupy/blob/bd467372871a14349eabd73b8115aab5ae9eeb85/cupy/_core/_routines_math.pyx#L956-L998).

Spark conforms to the STL C++ convention of always returning `double`. Therefore, we need to change the behavior of `Pow` and introduce a new struct `IntPow`:
1. `Pow` should always return `double` (it will fail if the requested output type is not `double`).
2. cuDF Python will dispatch to `IntPow`, which will implement exponentiation-by-squaring. This algorithm is already implemented in libcudf for [fixed-point powers](https://github.com/rapidsai/cudf/blob/97db4ba6c2287388a9859693df3f6f1a48bed7e6/cpp/include/cudf/fixed_point/fixed_point.hpp#L82-L98).

The only piece that still requires some investigation is the choice of conventions for `float` values, like `pow(float, float) -> (float or double?)` across the PyData, STL, and Spark ecosystems. It appears Pandas and C++ STL would return `float` but it looks like [Spark](https://scala-lang.org/api/3.x/scala/math.html#pow-44b)/[Java](https://docs.oracle.com/javase/10/docs/api/java/lang/Math.html#pow(double,double)) always upcast `pow` arguments to `double` and return `double`. We may want to allow `Pow` to perform its math in `float` precision instead of `double` to align with STL conventions -- currently libcudf casts `float` arguments to `double`, then downcasts the result to the common type of the inputs, `float`.

**Comment by bdice** [comment]
Issue #9346 is also relevant here, since that `ipow` implementation should be reused as a utility function.

**Comment by bdice** [comment]
This problem may also affect decimal types, since they use integers to represent the fixed point value. That should be verified before closing this issue.

**Comment by bdice** [comment]
> This problem may also affect decimal types, since they use integers to represent the fixed point value. That should be verified before closing this issue.

A separate issue affects decimal types. I filed #11636 and will close this issue.

**Comment by RAMitchell** [comment]
Is this fixed? I can do this with libcudf==24.12.00:
```C++
#include <cudf_test/column_wrapper.hpp>
#include <cudf_test/debug_utilities.hpp>
TEST(BinaryOpsTest, CudfPower){
  cudf::test::fixed_width_column_wrapper<int> a{{1, 2, 3, 4}};
  cudf::numeric_scalar<int> rhs{1, true};
  auto result = cudf::binary_operation(a, rhs, cudf::binary_operator::POW, cudf::data_type(cudf::type_id::INT32));
  cudf::test::print(result->view());
}
```
```
1,2,2,4
```

**Comment by bdice** [comment]
@RAMitchell Yes. Please use the `INT_POW` operator for exact powers of integer inputs. The `POW` operator always uses floating point math which is not exact when casting integral types to floating point values and back to integral types.

We retain the existing `POW` behavior of using floating-point types for the computation, because power computations can easily overflow an integer output. It's common to have `pow(int, int) -> double`. Perhaps we could specialize `pow(Integral, Integral) -> Integral` to use an integer power computation, but there's some surprise factor there because performance won't match.
